packages/app/src/cli/services/dev/extension/bundler.ts

```diff
@@ -207,3 +215,101 @@ export async function setupConfigWatcher({
+  app: AppInterface
+  stdout: Writable
+  stderr: Writable
+  signal: AbortSignal
+  token: string
+  apiKey: string
+  registrationId: string
+  unifiedDeployment: boolean
+}
+
+export async function setupFunctionWatcher({
+  extension,
+  app,
+  stdout,
+  stderr,
+  signal,
+  token,
+  apiKey,
+  registrationId,
+  unifiedDeployment,
+}: SetupFunctionWatcherOptions) {
+  const {default: chokidar} = await import('chokidar')
+
+  outputDebug(`Starting watcher for function extension ${extension.devUUID}`, stdout)
+  const watchPaths = getFunctionWatchPaths(extension as ExtensionInstance<FunctionConfigType>)
+  if (!watchPaths) {
+    outputWarn(
+      `Function extension ${extension.localIdentifier} does not have a watch path configured, draft version deployment is disabled.`,
+      stdout,
+    )
+    return
+  }
+
+  outputDebug(`Watching paths for function extension ${extension.localIdentifier}: ${watchPaths}`, stdout)
+  let buildController: AbortController | null
+
+  const functionWatcher = chokidar.watch(watchPaths).on('change', (path) => {
+    outputDebug(`Function extension file at path ${path} changed`, stdout)
+    if (buildController) {
+      // terminate any existing builds
+      buildController.abort()
+    }
+    buildController = new AbortController()
+    const buildSignal = buildController.signal
+    buildFunctionExtension(extension, {
+      app,
+      stdout,
+      stderr,
+      useTasks: false,
+      signal: buildSignal,
+    })
+      .then(() => {
+        if (!buildSignal.aborted) {
+          return updateExtensionDraft({extension, token, apiKey, registrationId, stderr, unifiedDeployment})
+        }
+      })
+      .catch((updateError: unknown) => {
+        outputWarn(`Error while deploying updated extension draft: ${JSON.stringify(updateError, null, 2)}`, stdout)
+      })
+      .finally(() => {
+        buildController = null
+      })
+  })
+
+  signal.addEventListener('abort', () => {
+    outputDebug(`Closing function file watching for extension with ID ${extension.devUUID}`, stdout)
+    functionWatcher
+      .close()
+      .then(() => {
+        outputDebug(`Function file watching closed for extension with ${extension.devUUID}`, stdout)
+      })
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      .catch((error: any) => {
+        outputDebug(
+          `Function file watching failed to close for extension with ${extension.devUUID}: ${error.message}`,
+          stderr,
+        )
+      })
+  })
+}
+
+export function getFunctionWatchPaths(extension: ExtensionInstance<FunctionConfigType>) {
+  if (!extension.isJavaScript && extension.watchPaths.length === 0) {
+    return null
+  }
+
+  const watchPaths: string[] = extension.watchPaths ?? []
+  if (extension.isJavaScript && watchPaths.length === 0) {
+    watchPaths.push(joinPath('src', '**', '*.js'))
+    watchPaths.push(joinPath('src', '**', '*.ts'))
+  }
+  watchPaths.push(joinPath('**', 'input*.graphql'))
```

Comment:
Maybe we can also detect if a function is a Rust function and add some defaults for that case as well, what do you think?